swarm/src/behaviour.rs

```diff
@@ -40,6 +41,47 @@ pub(crate) type THandlerInEvent<THandler> =
 pub(crate) type THandlerOutEvent<THandler> =
     <<THandler as IntoConnectionHandler>::Handler as ConnectionHandler>::OutEvent;
 
+/// Implementation of [`NetworkBehaviour`] that combines two underlying implementations.
+#[cfg(feature = "macros")]
+#[derive(Debug, Default, libp2p_swarm_derive::NetworkBehaviour)]
```

Comment:
Right now this will have the macro-generated `OutEvent` looking kinda like this: 
```rust
enum AndEvent<A, B> {
    First(A)
    Second(B)
}
```
What do you think about specifying `Either<A, B>` as out-event instead? Would also be more consistent with all of our other "combinators".

Reply:
Good point. Note @jxs that you can still achieve the above suggestion with the derive macro by using the `out_event` derive parameter, see https://docs.rs/libp2p-swarm/latest/libp2p_swarm/behaviour/trait.NetworkBehaviour.html#custom-networkbehaviour-with-the-derive-macro.

Reply:
Yeah I tried that while experimenting, but when you specify a custom `out_event` one has to impl `From` for each subvariant. In case of using `Either` the generated code is the following:
```Rust
    impl<A, B> crate::derive_prelude::NetworkBehaviour for And<A, B>
    where
        A: crate::derive_prelude::NetworkBehaviour,
        B: crate::derive_prelude::NetworkBehaviour,
        Either<A, B>: From<<A as crate::derive_prelude::NetworkBehaviour>::OutEvent>,
        Either<A, B>: From<<B as crate::derive_prelude::NetworkBehaviour>::OutEvent>,
```
which kinda doesn't make sense as it would require for us to know `A` and `B`, and means that specifying `out_event` for a derived `NetworkBehaviour` with generics is not supported now.  Thanks for referring that Elena. I am kind of out of ideas to allow `derive` to support custom `out_event` for generic structures. Do you ave any suggestion? Else I can add this to the doc.

Reply:
Ahh yes you're right. Using `Either` as out-event would only be possible if we manually derive `NetworkBehaviour`. I don't feel strongly about it, thus I'm fine with sticking with the current solution.

> I am kind of out of ideas to allow derive to support custom out_event for generic structures. Do you ave any suggestion?

I don't think its possible. We will always run into the issue of conflicting implementations if we try to auto-implement `From`, since A and B theoretically may have the same type.

Reply:
I think this is a point for not providing an `And` behaviour. If a user wants to group a set of `NetworkBehaviour`s together, they can do so themselves and provide a more meaningful name than `And`. Like, if they combine certain behaviours together than are only available to under a "legacy" feature-toggle, they can call this behaviour `Legacy` and the generated name will make more sense.

Plus, they can add as many behaviours as they want, not just two, they can decide on the visibility, they can decide on the names etc.

I really don't see the value in providing an `And` behaviour.